presto-main/src/main/java/com/facebook/presto/matching/MatchablesStore.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.matching;
+
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Streams;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class MatchablesStore<T extends Matchable>
+{
+    private final ListMultimap<Class, T> matchablesByClass;
+
+    private MatchablesStore(ListMultimap<Class, T> matchablesByClass)
+    {
+        this.matchablesByClass = ImmutableListMultimap.copyOf(matchablesByClass);
+    }
+
+    public Stream<T> getCandidates(Object object)
+    {
+        return Streams.stream(ancestors(object.getClass()))
+                .flatMap(clazz -> matchablesByClass.get(clazz).stream());
+    }
+
+    private static Iterator<Class> ancestors(Class clazz)
```

Comment:
I know it was here before.
But this method does not look good. It does not handle interfaces correctly.

More correct, but possibly less performant approach would be to implement `getCandidates` as iteration over keys of `matchablesByClass` 
and for each check if `key.isInstance(object)`